text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,366 @@
+would reduce the need for setting environment variables manually. Simply running
+`cargo test --coverage` would automatically run a build setting the `-C instrument-coverage`
+rustc flag only for the crates selected and set the `LLVM_PROFILE_FILE` environment
+variable to ensure each test run produces a unique `profraw` file. Once the tests have
+finished running, Cargo would leverage LLVM tooling to analyze profraw files and generate
+coverage reports for a Rust developer.
+
+This design does not break any existing usage of Rust or Cargo. This new feature would
+be strictly opt-in. A Rust developer would still be able to manually set the
+`-C instrument-coverage` rustc flag and instrument all binaries within the dependency
+graph. Since this is a Cargo specific feature, the Rust compiler will not need any updates.
+
+## Alternatives
+
+### Alternative 1: leave the existing feature
+
+Supporting this alternative would mean that no changes would be necessary to either Cargo
+or Rust. Getting instrumentation-based code coverge is still supported and would continue
+to work as it does today.
+
+The drawback for this option is that it would require setting the flag for all crates in
+the dependency graph, including upstream dependencies. This would also instrument all
+binaries and report on coverage for functions and branches that are not defined by the
+current crate with the potential of skewing coverage results.
+
+### Alternative 2: use a new manifest key instead of a CLI option
+
+Supporting this alternative would mean that changes would need to be made to the existing
+manifest format to possibly include a new section and/or key. A new `coverage` key could
+be added to the target section, `coverage = true`. This still has the added benefit of not
+requiring any changes to the Rust compiler itself and the feature could be scoped to Cargo only.
+
+The drawback for this option is that it could potentially add clutter to the `Cargo.toml`
+files. Adding this new section to every crate that a Rust developer wants to have instrumented
+would only add to all the data that is already contained within a toml file.
+
+### Alternative 3: use a RUSTC_WRAPPER program to selectively choose which crates to profile
+
+Supporting this alternative would mean that there wouldn't need to be any changes to Cargo
+at all.
+
+This would require creating a RUSTC_WRAPPER program specifically for selecting which crates to profile.
+This means more boiler plate code for each Rust developer that simply wants to profile their own crate.
+I believe the feature this RFC proposes would both be a cleaner solution long term and more in line
+with the Cargo workflow of potentially reading these kinds of behaviors from the `Cargo.toml` manifest
+file.
+
+### Alternative 4: use existing custom subcommands to run code coverage analysis
+
+Supporting this alternative would mean that there wouldn't need to be any changes to Cargo
```

Comment:
Given that these options exist today, I think the tone of the RFC should be more about 'moving an existing plugin from the ecosystem to core Cargo' rather than 'adding a new feature to Cargo'. Concretely, that means justifying why such a move would benefit a large enough subset of users to justify the increased maintenance burden, and what the benefits of core support rather than ecosystem support actually are. Furthermore, if you're proposing something new rather than something which already exists then you need to justify why the things which exist (or at least their API) are less good than the proposal.

I think you should also consider an alternative where we add features to Cargo which enable the ecosystem plugins to be as good as a core feature, i.e., could we add features which let cargo-llvm-cov be as good as the proposal in this RFC?